### PR TITLE
Update to cmake v3.17.0

### PR DESCRIPTION
**Pull request by trxcllnt**
Checklist
* [x] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
   name: cmake
   version: {{ version }}
 
 source:
   url: https://gitlab.kitware.com/cmake/cmake/-/archive/v{{ version }}/cmake-v{{ version }}.tar.bz2
-  sha256: 40e0dec6dc9e36820e001b8425aa4328a0b42f1915b14d68aee116e25c3d34df
+  sha256: 9791a8ff223ffbecc5a8985d4a623be8bff037881356f7147fa616e18c23f8c6
   patches:
     - patches/3.16.2/0001-find_-Add-debug-logging-infrastructure.patch
     - patches/3.16.2/0002-find_-Use-debug-logging-infrastructure.patch
```
Comment (jakirkham) [inline recipe/meta.yaml]: We may wind up needing to drop or alter these patches, but let's wait for CI to report.

Comment (jakirkham) [reply recipe/meta.yaml]: @mingwandroid, appears you added these, do you know if we still need them for 3.17.0?

Comment (jakirkham) [reply recipe/meta.yaml]: Yeah looks like these don't apply on CI. Probably got fixed upstream. Let's try dropping them 🙂

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,16 +1,12 @@
...
   name: cmake
   version: {{ version }}
 
 source:
   url: https://gitlab.kitware.com/cmake/cmake/-/archive/v{{ version }}/cmake-v{{ version }}.tar.bz2
-  sha256: 40e0dec6dc9e36820e001b8425aa4328a0b42f1915b14d68aee116e25c3d34df
-  patches:
-    - patches/3.16.2/0001-find_-Add-debug-logging-infrastructure.patch
-    - patches/3.16.2/0002-find_-Use-debug-logging-infrastructure.patch
-    - patches/3.16.2/0003-Add-more-debug-logging-to-cmFindCommon.patch
```
Comment (jakirkham) [inline recipe/meta.yaml]: Would also delete these files too 🙂

**Comment by jakirkham** [comment]
Restarted CI as it appeared something funky was going on.

**Comment by jakirkham** [comment]
I wonder if we are seeing some kind of `conda-build` bug. It appears to be checking if macOS is used in [the Linux build]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=135336&view=logs&jobId=e018f978-a0ce-53fb-9fb6-ffd586bfca01&j=e018f978-a0ce-53fb-9fb6-ffd586bfca01&t=5c69329b-d947-5705-66b5-610e0963c1aa ) (also [full CI text log]( https://github.com/conda-forge/cmake-feedstock/files/4377373/cmake_linux64.txt )).

```python
WARNING :: get_rpaths_raw()=['/home/conda/feedstock_root/build_artifacts/cmake_1585077223713/_build_env/lib:/home/conda/feedstock_root/build_artifacts/cmake_1585077223713/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib'] and patchelf=[['/home/conda/feedstock_root/build_artifacts/cmake_1585077223713/_build_env/lib', '/home/conda/feedstock_root/build_artifacts/cmake_1585077223713/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib']] disagree for /home/conda/feedstock_root/build_artifacts/cmake_1585077223713/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/cmake :: 
Warning: rpath /home/conda/feedstock_root/build_artifacts/cmake_1585077223713/_build_env/lib is outside prefix /home/conda/feedstock_root/build_artifacts/cmake_1585077223713/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place (removing it)
Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 474, in main
    execute(sys.argv[1:])
  File "/opt/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 465, in execute
    verify=args.verify, variants=args.variants)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/api.py", line 209, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 2868, in build_tree
    notest=notest,
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 2155, in build
    newly_built_packages = bundlers[pkg_type](output_d, m, env, stats)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1509, in bundle_conda
    files = post_process_files(metadata, initial_files)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1368, in post_process_files
    post_build(m, new_files, build_python=python)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/post.py", line 1174, in post_build
    post_process_shared_lib(m, f, prefix_files, host_prefix)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/post.py", line 1122, in post_process_shared_lib
    mk_relative_osx(path, host_prefix, m.config.build_prefix, files=files, rpaths=rpaths)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/post.py", line 399, in mk_relative_osx
    assert sys.platform == 'darwin'
AssertionError
```

**Comment by jakirkham** [comment]
FWIW that discussion has led to PR ( https://github.com/conda/conda-build/pull/3912 ).

**Comment by mingwandroid** [comment]
Do we know which file this is or how Linux may be creating a macOS binary?

**Comment by jakirkham** [comment]
It's `share/cmake-3.17/Modules/Internal/CPack/CPack.OSXScriptLauncher.in`. I think it is [already in the source distribution]( https://gitlab.kitware.com/cmake/cmake/-/blob/v3.17.0/Modules/Internal/CPack/CPack.OSXScriptLauncher.in ).

cc @chrisburr (in case I'm missing anything 🙂)

**Comment by mingwandroid** [comment]
Thanks, we should implement a way to explicitly disable RPATH fixup for specific files I think.

**Comment by jakirkham** [comment]
That would be helpful 🙂

What would you recommend we do in the interim?

Also do you have any thoughts on the patches above? We dropped them as we figured they got upstreamed, but please let us know if that is not the case and we should do something else.

**Comment by mingwandroid** [comment]
Well we need to build cmake so lgtm.

**Comment by chrisburr** [comment]
Well we need to downgrade conda-build for this to work. I suggested editing `.scripts/build_steps.sh` to replace `conda-build` with `'conda-build<3.18'`.

File: .scripts/build_steps.sh [hunk 3]
```diff
@@ -19,7 +19,7 @@ conda-build:
 
 CONDARC
 
-conda install --yes --quiet conda-forge-ci-setup=2 conda-build -c conda-forge
+conda install --yes --quiet conda-forge-ci-setup=2 "conda-build<3.18" -c conda-forge
```
Comment (chrisburr) [inline .scripts/build_steps.sh]: Sorry about that, it seems it's bugs all the way down! This version worked on `master`:
```suggestion
conda install --yes --quiet conda-forge-ci-setup=2 "conda-build=3.18.11" -c conda-forge
```
Taken from https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=121043&view=logs&j=1bf226d3-0e2f-52d8-fa93-7d9e633347b3&t=ec8d466d-e3b4-5115-4f06-222398f91e6c

**Comment by jakirkham** [comment]
Seeing a weird error on Windows. Maybe the package for `pkginfo` is broken?

```
ERROR conda.core.link:_execute(568): An error occurred while installing package 'conda-forge::pkginfo-1.5.0.1-py_0'.
CondaError: Cannot link a source that does not exist. C:\Miniconda\Scripts\conda.exe
Running `conda clean --packages` may resolve your problem.
Attempting to roll back.
```

Also here's the [log on CI]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=136172&view=logs&j=bcc8647a-ade0-513f-0af3-8dc720e6c7f2&t=695bc08b-8f56-5fad-5b7a-5e070472f8f1 ) and [raw text log for download]( https://github.com/conda-forge/cmake-feedstock/files/4390072/cmake_win64.txt ).

**Comment by jakirkham** [comment]
To fix the AppVeyor issue, we need to re-render it seems. However that would wipe out the changes here. So I requested the bot create PR ( https://github.com/conda-forge/cmake-feedstock/pull/117 ) with the re-render. Though of course that needed the fixes from here. 😂 So I wound up cherry-picking the fixes into that PR. Once that PR passes will merge and then merge the other update PR ( https://github.com/conda-forge/cmake-feedstock/pull/114 ) and this PR (finally shaving the yak 😉).

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render.

**Comment by jakirkham** [comment]
Thanks @trxcllnt for the PR and everyone for the reviews! 😄

**Pull request merged**